### PR TITLE
Start writing docs.

### PR DESCRIPTION
**Pull request by mkcor**
Hi @lagru,

Here I'm starting some minimal documentation as per #9.

I'm using the same structure and formats as scikit-image's so we can use the same machinery to build the docs, i.e., generated the rendered HTML pages. I guess we should deploy these somewhere?

I've noticed that

* the scikit-image docs live at https://scikit-image.org/docs/stable/ while
* the NumPy docs live at https://numpy.org/doc/stable/ (no -s).

Maybe it would be more consistent to have `doc/src/index.rst` instead of `doc/source/index.rst`?

### CODE REVIEW

**Comment by lagru** [comment]
Thanks for getting this going @mkcor!

I've been thinking about how to keep the effort for the documentation infrastructure low during early development. E.g. start with a Wiki on the Repo? I'm also considering to go "full markdown" for docstubs documentation via [MyST](https://myst-parser.readthedocs.io/en/latest/)? 

I'm curious what you think. 

> Maybe it would be more consistent to have doc/src/index.rst instead of doc/source/index.rst?

I've never really been a fan of the extra `doc/source/` folder / level in scikit-image and would love to avoid it here if we can.

**Comment by mkcor** [comment]
Hi @lagru!

> I've never really been a fan of the extra `doc/source/` folder / level in scikit-image and would love to avoid it here if we can.

I totally agree, `doc/index.rst` should be good enough!

> I've been thinking about how to keep the effort for the documentation infrastructure low during early development. E.g. start with a Wiki on the Repo? I'm also considering to go "full markdown" for docstubs documentation via [MyST](https://myst-parser.readthedocs.io/en/latest/)?

Nice, I like being minimalist. Good idea; let's keep the docs in the repo's wiki, at least for now. We can always move them later, and they'll be in Markdown already.

PS: I found out about MyST last year at JupyterCon and would love a pretext to try it out!

**Comment by mkcor** [comment]
@lagru would you mind creating the wiki for this repo? I personally don't have access to the 'Settings' tab... I guess that's where you can check the box to get the Wiki. Thanks!

**Comment by mkcor** [comment]
> would you mind creating the wiki for this repo?

Or, alternatively, give me permission to do so? Any update on this, @lagru? Thanks.

**Comment by lagru** [comment]
Oh, I didn't realize that you needed permissions. I'll grant those, but maybe its easiest to just commit the markdown files in the main repo instead of taking the detour via wiki. We can worry about the plumbing for a HTML documentation later. 

That way your contributions are more visible also, stay preserved and we have a proper review process. :)

**Comment by mkcor** [comment]
> maybe its easiest to just commit the markdown files in the main repo

You're right.
 
> That way your contributions are more visible also, stay preserved and we have a proper review process. :)

Very valid point! Good, I'll update this PR then, and mark it as ready for review.

**Comment by mkcor** [comment]
@lagru the action is happening at 699c509f3552cb85287e0158b11d520be21c3996 :shipit:

File: doc/index.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# docstub user guide
+
+**Version**:
```
Comment (mkcor) [inline doc/index.md:3]: I don't know how to insert a variable in here...

Comment (mkcor) [reply doc/index.md:3]: For now, we could hard-write it.

File: doc/index.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+def example_metric(
+    image0: ArrayLike,
+    image1: ArrayLike,
+    sigma: float = ...,
+    method: Literal["standard", "modified"] = ...,
+) -> NDArray[float]
+```
+
+As you can see, it is a typed function signature, where types are read from
+the well-enough written docstring.
```
Comment (mkcor) [inline doc/index.md:67]:
```suggestion
the (well-enough) written docstring.
```
Maybe?

File: doc/index.md (hunk 1 above)
Comment (lagru) [reply doc/index.md:3]: Honestly, I think we can skip this for now. It should be pretty clear that the user guide reflects the current version of the repo. Longterm this should end up in a website with a version switcher anyway. :)

**Comment by lagru** [comment]
@mkcor, do you mind if I go through this PR and tweak and update as you go? I think that might be the fastest approach to fleshing out this guide some more. After that we can have another round of iteration. I'd be especially thankful for your eyes which are less familiar with the inner workings!

**Comment by mkcor** [comment]
I don't mind at all, please go ahead!

**Comment by lagru** [comment]
@mkcor I've started to iterate on this heavily since there's a lot to tell which currently lives only in my head. Please feel very welcome to come down on me and on any section that isn't clear to you! :pray: 

Currently, I'm focusing on being clear and getting the main ideas across rather than perfect wording. Which is hard enough, if I let it, my brain defaults to going for "95% done" or nothing. :sweat_smile:

File: doc/user_guide.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+def example_metric(image, *, mask=None, sigma=1.0, method='standard'):
+    """Pretend to calculate a local metric between two images.
+
+    Parameters
+    ----------
+    image : array-like
+        First image.
+    mask : array of dtype uint8, optional
+        Second image.
+    sigma : float or Iterable of float, optional
```
Comment (OriolAbril) [inline doc/user_guide.md:40]:
```suggestion
    sigma : float or Iterable of float, default = 1
```

This is mostly personal preference but I think makes the example closer to most common use within the community. According to numpydoc:

> Optional keyword parameters have default values, which are displayed as part of the function signature. They can also be detailed in the description [...] or as part of the type, instead of optional. If the default value would not be used as a value, optional is preferred.

To me from this is is clear that optional and default shouldn't be used together (next comment) and that when the default value is not the one being used (which I would assume to be the case of `mask` as in `None` means no masking and not `image[mask]` with `mask=None` to add a new axis to the array) it is preferred to use `, optional`.

That would mean that when `sigma=1` is the default there is no preference for `default 1` or `optional` but in practice from what I have seen, `default 1` (with equal, space or colon) is preferred to `optional`.

File: doc/user_guide.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        Sigma value for each dimension in `image`. A single value is broadcast
+        to all dimensions.
+    method : {'standard', 'modified'}, optional, default = 'standard'
```
Comment (OriolAbril) [inline doc/user_guide.md:43]:
```suggestion
    method : {'standard', 'modified'}, default = 'standard'
```

File: doc/typing_syntax.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+Since docstrings are also used to generate documentation with Sphinx, you may want to use [restructuredText roles](https://docutils.sourceforge.io/docs/ref/rst/roles.html) in your type annotations.
+Docstub allows for this anywhere where a qualified name can be used.
+
+| Docstring type       | Python type annotation |
+|----------------------|------------------------|
+| `` `X` ``            | `X`                    |
+| ``:ref:`X` ``        | `X`                    |
+| ``:class:`Y.X` ``    | `Y.X`                  |
+| ``:py:class:`Y.X` `` | `Y.X`                  |
```
Comment (OriolAbril) [inline doc/typing_syntax.md:124]: This is very nice! I have actually used it already but didn't properly realize it was happening. I will test it out for potential edge cases (tildes, explicit text, ommited imports) and report back. Matplotlib has a lot of them for example: https://matplotlib.org/devdocs/devel/document.html#reference-types

File: doc/user_guide.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    """
+    pass
+```
+
+<!--- end example.py --->
+
+Feeding this input to docstub with
+
+```shell
+docstub simple_script.py
```
Comment (OriolAbril) [inline doc/user_guide.md:58]:
```suggestion
docstub example.py
```

**Review by OriolAbril — COMMENTED** [review]
I have been managing to use without problem so far but I always enjoy reading docs. I commented on a couple nit-picks I noticed while going over the pages, feel free to use or ignore.
